config_train.txt
--noamp
--anneal_rate=5e-05
--noapply_gradient_penalty

--aug_prob=0.0
--aug_types=cutout
--aug_types=translation
--batch_size=6
--c=0.01
--checkpoint=./results/vqvae_stage1/model-24000.pt
--checkpoint_every_steps=2000
--d_fmap_max=256
--data=/mnt/ssd1/instagram_global/,/home/theblackcat102/img_corpus/m500/instagram_global/,/home/theblackcat102/img_corpus/unsplash_caption/
--dataset_aug_prob=0.1
--dec_attn_res_layers=16
--nodetach_z

--dis_step=1
--dis_type=patch
--disc_output_size=6
--disc_weight=1.0
--discriminator_iter_start=2000
--downsample=32
--nodual_contrast_loss

--fmap_max=512
--nofreq_chan_attn
--gp_weight=10.0
--gradient_accumulate_every=2
--nogreyscale
--image_size=512
--init_temp=2.5
--latent_dim=512
--learning_rate=0.0001
--min_temp=0.5
--name=vqgan_stage2.1
--num_gpus=1
--num_train_steps=2000000
--num_workers=10
--optimizer=adam
--perceptual_weight=-1.0
--recon_only
--recon_weight=1.0
--sample_every_steps=1000
--sample_grid_size=5
--smooth_l1_loss
--notransparent
--ttur_mult=1.0
--update_temp=100
--noviz
--vocab_size=18384
--warmup_steps=1000
--nohelp
--nohelpfull
--nohelpshort
--nohelpxml
--noonly_check_args
--nopdb
--nopdb_post_mortem
--norun_with_pdb
--norun_with_profiling
--use_cprofile_for_profiling
--noalsologtostderr
--log_dir=
--logger_levels=
--nologtostderr
--showprefixforinfo
--stderrthreshold=fatal
--verbosity=0


config_train_docker.txt
--noamp
--anneal_rate=5e-05
--noapply_gradient_penalty

--aug_prob=0.0
--aug_types=cutout
--aug_types=translation
--batch_size=6
--c=0.01
--checkpoint=./results/vqvae_stage1/model-24000.pt
--checkpoint_every_steps=2000
--d_fmap_max=256
--data=/ssd1/instagram_global/,/data/m500/instagram_global/,/data/unsplash_caption/
--dataset_aug_prob=0.1
--dec_attn_res_layers=16
--nodetach_z

--dis_step=1
--dis_type=patch
--disc_output_size=6
--disc_weight=1.0
--discriminator_iter_start=2000
--downsample=32
--nodual_contrast_loss

--fmap_max=512
--nofreq_chan_attn
--gp_weight=10.0
--gradient_accumulate_every=2
--nogreyscale
--image_size=512
--init_temp=2.5
--latent_dim=512
--learning_rate=0.0001
--min_temp=0.5
--name=vqgan_stage2.1
--num_gpus=1
--num_train_steps=2000000
--num_workers=10
--optimizer=adam
--perceptual_weight=-1.0
--recon_only
--recon_weight=1.0
--sample_every_steps=1000
--sample_grid_size=5
--smooth_l1_loss
--notransparent
--ttur_mult=1.0
--update_temp=100
--noviz
--vocab_size=18384
--warmup_steps=1000
--nohelp
--nohelpfull
--nohelpshort
--nohelpxml
--noonly_check_args
--nopdb
--nopdb_post_mortem
--norun_with_pdb
--norun_with_profiling
--use_cprofile_for_profiling
--noalsologtostderr
--log_dir=
--logger_levels=
--nologtostderr
--showprefixforinfo
--stderrthreshold=fatal
--verbosity=0


lightweight_gan/cli.py
import os
import fire
import random
from retry.api import retry_call
from tqdm import tqdm
from datetime import datetime
from functools import wraps
from lightweight_gan import Trainer, NanException
from lightweight_gan.diff_augment_test import DiffAugmentTest

import torch
torch.backends.cudnn.benchmark = True
import torch.multiprocessing as mp
import torch.distributed as dist

import numpy as np

def exists(val):
    return val is not None

def default(val, d):
    return val if exists(val) else d

def cast_list(el):
    return el if isinstance(el, list) else [el]

def timestamped_filename(prefix = 'generated-'):
    now = datetime.now()
    timestamp = now.strftime("%m-%d-%Y_%H-%M-%S")
    return f'{prefix}{timestamp}'

def set_seed(seed):
    torch.manual_seed(seed)
    torch.backends.cudnn.deterministic = True
    torch.backends.cudnn.benchmark = False
    np.random.seed(seed)
    random.seed(seed)

def run_training(rank, world_size, model_args, data, load_from, new, num_train_steps, name, seed):
    is_main = rank == 0
    is_ddp = world_size > 1

    if is_ddp:
        set_seed(seed)
        os.environ['MASTER_ADDR'] = 'localhost'
        os.environ['MASTER_PORT'] = '12355'
        dist.init_process_group('nccl', rank=rank, world_size=world_size)

        print(f"{rank + 1}/{world_size} process initialized.")

    model_args.update(
        is_ddp = is_ddp,
        rank = rank,
        world_size = world_size
    )

    model = Trainer(**model_args)

    if not new:
        model.load(load_from)
    else:
        model.clear()

    model.set_data_src(data)

    progress_bar = tqdm(initial = model.steps, total = num_train_steps, mininterval=10., desc=f'{name}<{data}>', dynamic_ncols=True)
    while model.steps < num_train_steps:
        retry_call(model.train, tries=3, exceptions=NanException)
        progress_bar.n = model.steps
        progress_bar.refresh()
        if is_main and model.steps % 50 == 0:
            model.print_log()

    model.save(model.checkpoint_num)

    if is_ddp:
        dist.destroy_process_group()

def train_from_folder(
    data = './data',
    results_dir = './results',
    models_dir = './models',
    name = 'default',
    new = False,
    load_from = -1,
    image_size = 256,
    optimizer = 'adam',
    fmap_max = 512,
    transparent = False,
    greyscale = False,
    batch_size = 10,
    gradient_accumulate_every = 4,
    num_train_steps = 150000,
    learning_rate = 2e-4,
    save_every = 1000,
    evaluate_every = 1000,
    generate = False,
    generate_types = ['default', 'ema'],
    generate_interpolation = False,
    aug_test = False,
    aug_prob=None,
    aug_types=['cutout', 'translation'],
    dataset_aug_prob=0.,
    attn_res_layers = [32],
    freq_chan_attn = False,
    disc_output_size = 1,
    dual_contrast_loss = False,
    antialias = False,
    interpolation_num_steps = 100,
    save_frames = False,
    num_image_tiles = None,
    num_workers = None,
    multi_gpus = False,
    calculate_fid_every = None,
    calculate_fid_num_images = 12800,
    clear_fid_cache = False,
    fid_cache = None,
    seed = 42,
    amp = False,
    show_progress = False,
):
    num_image_tiles = default(num_image_tiles, 4 if image_size > 512 else 8)

    model_args = dict(
        name = name,
        results_dir = results_dir,
        models_dir = models_dir,
        batch_size = batch_size,
        gradient_accumulate_every = gradient_accumulate_every,
        attn_res_layers = cast_list(attn_res_layers),
        freq_chan_attn = freq_chan_attn,
        disc_output_size = disc_output_size,
        dual_contrast_loss = dual_contrast_loss,
        antialias = antialias,
        image_size = image_size,
        num_image_tiles = num_image_tiles,
        optimizer = optimizer,
        num_workers = num_workers,
        fmap_max = fmap_max,
        transparent = transparent,
        greyscale = greyscale,
        lr = learning_rate,
        save_every = save_every,
        evaluate_every = evaluate_every,
        aug_prob = aug_prob,
        aug_types = cast_list(aug_types),
        dataset_aug_prob = dataset_aug_prob,
        calculate_fid_every = calculate_fid_every,
        calculate_fid_num_images = calculate_fid_num_images,
        clear_fid_cache = clear_fid_cache,
        fid_cache = fid_cache,
        amp = amp
    )

    if generate:
        model = Trainer(**model_args)
        model.load(load_from)
        samples_name = timestamped_filename()
        checkpoint = model.checkpoint_num
        dir_result = model.generate(samples_name, num_image_tiles, checkpoint, generate_types)
        print(f'sample images generated at {dir_result}')
        return

    if generate_interpolation:
        model = Trainer(**model_args)
        model.load(load_from)
        samples_name = timestamped_filename()
        model.generate_interpolation(samples_name, num_image_tiles, num_steps = interpolation_num_steps, save_frames = save_frames)
        print(f'interpolation generated at {results_dir}/{name}/{samples_name}')
        return

    if show_progress:
        model = Trainer(**model_args)
        model.show_progress(num_images=num_image_tiles, types=generate_types)
        return

    if aug_test:
        DiffAugmentTest(data=data, image_size=image_size, batch_size=batch_size, types=aug_types, nrow=num_image_tiles)
        return

    world_size = torch.cuda.device_count()

    if world_size == 1 or not multi_gpus:
        run_training(0, 1, model_args, data, load_from, new, num_train_steps, name, seed)
        return

    mp.spawn(run_training,
        args=(world_size, model_args, data, load_from, new, num_train_steps, name, seed),
        nprocs=world_size,
        join=True)

def main():
    fire.Fire(train_from_folder)
